include/envoy/registry/registry.h

```diff
@@ -132,11 +132,20 @@ template <class Base> class FactoryRegistry {
     return *factories;
   }
 
-  static void registerFactory(Base& factory, absl::string_view name) {
+  static std::vector<std::string>& deprecatedFactoryNames() {
+    static auto* deprecated_factory_names = new std::vector<std::string>();
+    return *deprecated_factory_names;
+  }
+
+  static void registerFactory(Base& factory, absl::string_view name, bool is_deprecated = false) {
     auto result = factories().emplace(std::make_pair(name, &factory));
     if (!result.second) {
       throw EnvoyException(fmt::format("Double registration for name: '{}'", factory.name()));
     }
+
+    if (is_deprecated) {
+      deprecatedFactoryNames().emplace_back(name);
```

Comment:
The only slight concern I have here is that we may end up with a situation one day where the same name is used for different extensions in a deprecated way. I'm thinking in particular about the transport sockets, which have the same names used for upstream and downstream; it's conceivable (but not likely) that we deprecate one but not the other..

Reply:
this is already in `FactoryRegistry<Base>` so it shouldn't cause conflict between upstream/downstream itself. But it might cause somewhere else (e.g. stat).